File: leancomfy/haskell/Lattice/Utils/Validation.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE DeriveFunctor #-}

{-|
Module      : Lattice.Utils.Validation
Description : Runtime input validation with Result types
Copyright   : (c) Lattice, 2026
License     : MIT

Validates input at runtime boundaries (AI commands, file imports, user input).
All validators return explicit success/failure results.

Source: leancomfy/lean/Lattice/Utils/Validation.lean
-}

module Lattice.Utils.Validation
  ( -- * Result Type
    ValidationResult(..)
  , ok
  , fail'
  , isOk
  , getOr
    -- * Primitive Validators
  , StringOptions(..)
  , validateString
  , NumberOptions(..)
  , validateFiniteNumber
  , validateInteger
  , validateBoolean
  , validateEnum
    -- * Numeric Type Validators
  , validatePositive
  , validateNonNegative
  , validateUnit
  , validatePercentage
  , validateFrameNumber
    -- * Array Validators
  , validateArray
  , validateNumberArray
    -- * Vector Validators
  , validateVec2
  , validateVec3
  , validateVec4
    -- * Color Validators
  , validateRGB
  , validateRGBA
    -- * Optional/Default Helpers
  , optional
  , withDefault
    -- * Composition Helpers
  , validateAll
  , andThen
  , both
  ) where

import Prelude hiding (fail)
import Data.Text (Text)
import qualified Data.Text as T
import Lattice.Primitives

--------------------------------------------------------------------------------
-- Result Type
--------------------------------------------------------------------------------

-- | Validation result - either success with typed value, or failure with error
data ValidationResult a
  = Ok !a
  | Fail !Text
  deriving (Eq, Show, Functor)

instance Applicative ValidationResult where
  pure = Ok
  Ok f <*> Ok x = Ok (f x)
  Fail e <*> _ = Fail e
  _ <*> Fail e = Fail e

instance Monad ValidationResult where
  Ok x >>= f = f x
  Fail e >>= _ = Fail e

-- | Create success result
ok :: a -> ValidationResult a
ok = Ok

-- | Create failure result
fail' :: Text -> ValidationResult a
fail' = Fail

-- | Check if result is success
isOk :: ValidationResult a -> Bool
isOk (Ok _) = True
isOk (Fail _) = False

-- | Get value with default fallback
getOr :: a -> ValidationResult a -> a
getOr _ (Ok v) = v
getOr d (Fail _) = d

--------------------------------------------------------------------------------
-- Primitive Validators
--------------------------------------------------------------------------------

-- | Options for string validation
data StringOptions = StringOptions
  { soMinLength :: !(Maybe Int)
  , soMaxLength :: !(Maybe Int)
  , soAllowEmpty :: !Bool
  } deriving (Eq, Show)

defaultStringOptions :: StringOptions
defaultStringOptions = StringOptions Nothing Nothing False

-- | Validate non-empty string
validateString :: Text -> Text -> StringOptions -> ValidationResult NonEmptyString
validateString value name opts
  | not (soAllowEmpty opts) && T.null value =
      Fail (name <> " cannot be empty")
  | Just minLen <- soMinLength opts, T.length value < minLen =
      Fail (name <> " must be at least " <> T.pack (show minLen) <> " characters")
  | Just maxLen <- soMaxLength opts, T.length value > maxLen =
      Fail (name <> " must be at most " <> T.pack (show maxLen) <> " characters")
  | T.null value = Fail (name <> " cannot be empty")
  | otherwise = Ok (NonEmptyString value)

-- | Options for numeric validation
data NumberOptions = NumberOptions
  { noMin :: !(Maybe Double)
  , noMax :: !(Maybe Double)
  } deriving (Eq, Show)

defaultNumberOptions :: NumberOptions
defaultNumberOptions = NumberOptions Nothing Nothing

-- | Check if Double is finite
isFiniteDouble :: Double -> Bool
isFiniteDouble x = not (isNaN x) && not (isInfinite x)

-- | Validate finite number
validateFiniteNumber :: Double -> Text -> NumberOptions -> ValidationResult FiniteFloat
validateFiniteNumber value name opts
  | not (isFiniteDouble value) =
      Fail (name <> " must be a finite number")
  | Just minVal <- noMin opts, value < minVal =
      Fail (name <> " must be >= " <> T.pack (show minVal) <> ", got " <> T.pack (show value))
  | Just maxVal <- noMax opts, value > maxVal =
      Fail (name <> " must be <= " <> T.pack (show maxVal) <> ", got " <> T.pack (show value))
  | otherwise = Ok (FiniteFloat value)

-- | Validate integer
validateInteger :: Double -> Text -> NumberOptions -> ValidationResult Int
validateInteger value name opts = do
  _ <- validateFiniteNumber value name opts
  let intVal = truncate value :: Int
  if fromIntegral intVal == value
    then Ok intVal
    else Fail (name <> " must be an integer, got " <> T.pack (show value))

-- | Validate boolean (always succeeds)
validateBoolean :: Bool -> Text -> ValidationResult Bool
validateBoolean = const . Ok

-- | Validate value is in allowed list
validateEnum :: (Eq a, Show a) => a -> Text -> [a] -> ValidationResult a
validateEnum value name allowed
  | value `elem` allowed = Ok value
  | otherwise = Fail (name <> " must be one of allowed values")

--------------------------------------------------------------------------------
-- Numeric Type Validators
--------------------------------------------------------------------------------

-- | Validate positive number (> 0)
validatePositive :: Double -> Text -> ValidationResult PositiveFloat
validatePositive value name
  | not (isFiniteDouble value) = Fail (name <> " must be a finite number")
  | value <= 0 = Fail (name <> " must be positive, got " <> T.pack (show value))
  | otherwise = Ok (PositiveFloat value)

-- | Validate non-negative number (>= 0)
validateNonNegative :: Double -> Text -> ValidationResult NonNegativeFloat
validateNonNegative value name
  | not (isFiniteDouble value) = Fail (name <> " must be a finite number")
  | value < 0 = Fail (name <> " must be non-negative, got " <> T.pack (show value))
  | otherwise = Ok (NonNegativeFloat value)

-- | Validate unit float (0 to 1)
validateUnit :: Double -> Text -> ValidationResult UnitFloat
validateUnit value name
  | not (isFiniteDouble value) = Fail (name <> " must be a finite number")
  | value < 0 = Fail (name <> " must be >= 0, got " <> T.pack (show value))
  | value > 1 = Fail (name <> " must be <= 1, got " <> T.pack (show value))
  | otherwise = Ok (UnitFloat value)

-- | Validate percentage (0 to 100)
validatePercentage :: Double -> Text -> ValidationResult Percentage
validatePercentage value name
  | not (isFiniteDouble value) = Fail (name <> " must be a finite number")
  | value < 0 = Fail (name <> " must be >= 0, got " <> T.pack (show value))
  | value > 100 = Fail (name <> " must be <= 100, got " <> T.pack (show value))
  | otherwise = Ok (Percentage value)

-- | Validate frame number (non-negative integer)
validateFrameNumber :: Double -> Text -> ValidationResult FrameNumber
validateFrameNumber value name
  | not (isFiniteDouble value) = Fail (name <> " must be a finite number")
  | value < 0 = Fail (name <> " must be non-negative, got " <> T.pack (show value))
  | otherwise =
      let intVal = truncate value :: Int
      in if fromIntegral intVal == value && intVal >= 0
         then Ok (FrameNumber (fromIntegral intVal))
         else Fail (name <> " must be a non-negative integer, got " <> T.pack (show value))

--------------------------------------------------------------------------------
-- Array Validators
--------------------------------------------------------------------------------

-- | Validate array with item validator
validateArray :: [Double] -> Text -> (Double -> Text -> ValidationResult a) -> ValidationResult [a]
validateArray values name itemValidator = go values 0 []
  where
    go [] _ acc = Ok (reverse acc)
    go (x:xs) idx acc =
      case itemValidator x (name <> "[" <> T.pack (show idx) <> "]") of
        Ok v -> go xs (idx + 1) (v : acc)
        Fail e -> Fail e

-- | Validate array of finite numbers
validateNumberArray :: [Double] -> Text -> NumberOptions -> ValidationResult [FiniteFloat]
validateNumberArray values name opts =
  validateArray values name (\v n -> validateFiniteNumber v n opts)

--------------------------------------------------------------------------------
-- Vector Validators
--------------------------------------------------------------------------------

-- | Validate Vec2
validateVec2 :: Double -> Double -> Text -> ValidationResult Vec2
validateVec2 x y name = do
  vx <- validateFiniteNumber x (name <> ".x") defaultNumberOptions
  vy <- validateFiniteNumber y (name <> ".y") defaultNumberOptions
  pure (Vec2 vx vy)

-- | Validate Vec3
validateVec3 :: Double -> Double -> Double -> Text -> ValidationResult Vec3
validateVec3 x y z name = do
  vx <- validateFiniteNumber x (name <> ".x") defaultNumberOptions
  vy <- validateFiniteNumber y (name <> ".y") defaultNumberOptions
  vz <- validateFiniteNumber z (name <> ".z") defaultNumberOptions
  pure (Vec3 vx vy vz)

-- | Validate Vec4
validateVec4 :: Double -> Double -> Double -> Double -> Text -> ValidationResult Vec4
validateVec4 x y z w name = do
  vx <- validateFiniteNumber x (name <> ".x") defaultNumberOptions
  vy <- validateFiniteNumber y (name <> ".y") defaultNumberOptions
  vz <- validateFiniteNumber z (name <> ".z") defaultNumberOptions
  vw <- validateFiniteNumber w (name <> ".w") defaultNumberOptions
  pure (Vec4 vx vy vz vw)

--------------------------------------------------------------------------------
-- Color Validators
--------------------------------------------------------------------------------

-- | Validate RGB color (0-255 each)
validateRGB :: Double -> Double -> Double -> Text -> ValidationResult RGB
validateRGB r g b name = do
  let colorOpts = NumberOptions (Just 0) (Just 255)
  vr <- validateFiniteNumber r (name <> ".r") colorOpts
  vg <- validateFiniteNumber g (name <> ".g") colorOpts
  vb <- validateFiniteNumber b (name <> ".b") colorOpts
  pure (RGB vr vg vb)

-- | Validate RGBA color
validateRGBA :: Double -> Double -> Double -> Double -> Text -> ValidationResult RGBA
validateRGBA r g b a name = do
  let colorOpts = NumberOptions (Just 0) (Just 255)
  vr <- validateFiniteNumber r (name <> ".r") colorOpts
  vg <- validateFiniteNumber g (name <> ".g") colorOpts
  vb <- validateFiniteNumber b (name <> ".b") colorOpts
  va <- validateUnit a (name <> ".a")
  pure (RGBA vr vg vb va)

--------------------------------------------------------------------------------
-- Optional/Default Helpers
--------------------------------------------------------------------------------

-- | Make validation optional
optional :: (Double -> Text -> ValidationResult a) -> Maybe Double -> Text -> ValidationResult (Maybe a)
optional _ Nothing _ = Ok Nothing
optional validator (Just v) name = Just <$> validator v name

-- | Provide default if validation fails
withDefault :: (Double -> Text -> ValidationResult a) -> a -> Double -> Text -> ValidationResult a
withDefault validator def v name =
  case validator v name of
    Ok x -> Ok x
    Fail _ -> Ok def

--------------------------------------------------------------------------------
-- Composition Helpers
--------------------------------------------------------------------------------

-- | Collect all errors from multiple validations
validateAll :: [ValidationResult ()] -> ValidationResult ()
validateAll validations =
  let errors = [e | Fail e <- validations]
  in if null errors
     then Ok ()
     else Fail (T.intercalate "; " errors)

-- | Chain two validations
andThen :: ValidationResult a -> (a -> ValidationResult b) -> ValidationResult b
andThen = (>>=)

-- | Combine two validations into a pair
both :: ValidationResult a -> ValidationResult b -> ValidationResult (a, b)
both (Ok a) (Ok b) = Ok (a, b)
both (Fail e) _ = Fail e
both _ (Fail e) = Fail e

File: comfyui-edition/ui/src/utils/validation.ts
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
//                                          // input // validation // utilities
// ════════════════════════════════════════════════════════════════════════════
// Runtime validation for input boundaries (AI commands, file imports, user input)
// These functions ACTUALLY CHECK values at runtime, not just assert types.
// ════════════════════════════════════════════════════════════════════════════

import type { JSONValue } from "@/types/dataAsset";

/**
 * All possible JavaScript values that can be validated at runtime
 * Used as input type for validators (replaces unknown)
 */
type RuntimeValue = string | number | boolean | object | null | undefined | bigint | symbol;

/**
 * Validation result - either success with typed value, or failure with error message
 */
export type ValidationResult<T> =
  | { ok: true; value: T }
  | { ok: false; error: string };

/**
 * Helper to create success result
 */
export function ok<T>(value: T): ValidationResult<T> {
  return { ok: true, value };
}

/**
 * Helper to create failure result
 */
export function fail<T>(error: string): ValidationResult<T> {
  return { ok: false, error };
}

// ════════════════════════════════════════════════════════════════════════════
//                                                   // primitive // validators
// ════════════════════════════════════════════════════════════════════════════

/**
 * Validate that value is a string
 */
export function validateString(
  value: RuntimeValue,
  name: string,
  options: { minLength?: number; maxLength?: number; allowEmpty?: boolean } = {},
): ValidationResult<string> {
  if (typeof value !== "string") {
    return fail(`${name} must be a string, got ${typeof value}`);
  }
  if (!options.allowEmpty && value.length === 0) {
    return fail(`${name} cannot be empty`);
  }
  if (options.minLength !== undefined && value.length < options.minLength) {
    return fail(`${name} must be at least ${options.minLength} characters`);
  }
  if (options.maxLength !== undefined && value.length > options.maxLength) {
    return fail(`${name} must be at most ${options.maxLength} characters`);
  }
  return ok(value);
}

/**
 * Validate that value is a finite number (not NaN, not Infinity)
 */
export function validateFiniteNumber(
  value: RuntimeValue,
  name: string,
  options: { min?: number; max?: number; allowNaN?: boolean } = {},
): ValidationResult<number> {
  const num = typeof value === "number" ? value : Number(value);

  if (Number.isNaN(num)) {
    if (options.allowNaN) return ok(num);
    return fail(`${name} must be a valid number, got NaN`);
  }

  if (!Number.isFinite(num)) {
    return fail(`${name} must be a finite number, got ${String(value)}`);
  }

  if (options.min !== undefined && num < options.min) {
    return fail(`${name} must be >= ${options.min}, got ${num}`);
  }

  if (options.max !== undefined && num > options.max) {
    return fail(`${name} must be <= ${options.max}, got ${num}`);
  }

  return ok(num);
}

/**
 * Validate that value is an integer
 */
export function validateInteger(
  value: RuntimeValue,
  name: string,
  options: { min?: number; max?: number } = {},
): ValidationResult<number> {
  const result = validateFiniteNumber(value, name, options);
  if (!result.ok) return result;

  if (!Number.isInteger(result.value)) {
    return fail(`${name} must be an integer, got ${result.value}`);
  }

  return result;
}

/**
 * Validate that value is a boolean
 */
export function validateBoolean(
  value: RuntimeValue,
  name: string,
): ValidationResult<boolean> {
  if (typeof value !== "boolean") {
    return fail(`${name} must be a boolean, got ${typeof value}`);
  }
  return ok(value);
}

/**
 * Validate that value is one of allowed values
 */
export function validateEnum<T extends string>(
  value: RuntimeValue,
  name: string,
  allowed: readonly T[],
): ValidationResult<T> {
  if (typeof value !== "string") {
    return fail(`${name} must be a string, got ${typeof value}`);
  }
  if (!allowed.includes(value as T)) {
    return fail(`${name} must be one of [${allowed.join(", ")}], got "${value}"`);
  }
  return ok(value as T);
}

// ════════════════════════════════════════════════════════════════════════════
//                                                       // array // validators
// ════════════════════════════════════════════════════════════════════════════

/**
 * Validate that value is an array
 */
export function validateArray<T extends JSONValue>(
  value: RuntimeValue,
  name: string,
  itemValidator: (item: JSONValue, index: number) => ValidationResult<T>,
  options: { minLength?: number; maxLength?: number } = {},
): ValidationResult<T[]> {
  if (!Array.isArray(value)) {
    return fail(`${name} must be an array, got ${typeof value}`);
  }

  if (options.minLength !== undefined && value.length < options.minLength) {
    return fail(`${name} must have at least ${options.minLength} items`);
  }

  if (options.maxLength !== undefined && value.length > options.maxLength) {
    return fail(`${name} must have at most ${options.maxLength} items`);
  }

  const result: T[] = [];
  for (let i = 0; i < value.length; i++) {
    const itemValue = value[i] as JSONValue;
    const itemResult = itemValidator(itemValue, i);
    if (!itemResult.ok) {
      return fail(`${name}[${i}]: ${itemResult.error}`);
    }
    result.push(itemResult.value);
  }

  return ok(result);
}

/**
 * Validate array of finite numbers
 */
export function validateNumberArray(
  value: RuntimeValue,
  name: string,
  options: {
    minLength?: number;
    maxLength?: number;
    min?: number;
    max?: number;
  } = {},
): ValidationResult<number[]> {
  return validateArray(
    value,
    name,
    (item, i) => {
      const itemValue = item as RuntimeValue;
      return validateFiniteNumber(itemValue, `item ${i}`, {
        min: options.min,
        max: options.max,
      });
    },
    { minLength: options.minLength, maxLength: options.maxLength },
  );
}

// ════════════════════════════════════════════════════════════════════════════
//                                                      // object // validators
// ════════════════════════════════════════════════════════════════════════════

/**
 * Generic object type for validation
 */
export interface ValidatedObject {
  [key: string]: JSONValue;
}

/**
 * Validate that value is a non-null object
 */
export function validateObject(
  value: RuntimeValue,
  name: string,
): ValidationResult<ValidatedObject> {
  if (typeof value !== "object" || value === null) {
    return fail(
      `${name} must be an object, got ${value === null ? "null" : typeof value}`,
    );
  }
  return ok(value as ValidatedObject);
}

/**
 * Validate a Vec2 (x, y coordinates)
 */
export function validateVec2(
  value: RuntimeValue,
  name: string,
): ValidationResult<{ x: number; y: number }> {
  const objResult = validateObject(value, name);
  if (!objResult.ok) return objResult;

  const obj = objResult.value;
  const xValue = obj.x as RuntimeValue;
  const x = validateFiniteNumber(xValue, `${name}.x`);
  if (!x.ok) return x;

  const yValue = obj.y as RuntimeValue;
  const y = validateFiniteNumber(yValue, `${name}.y`);
  if (!y.ok) return y;

  return ok({ x: x.value, y: y.value });
}

/**
 * Validate a Vec3 (x, y, z coordinates)
 */
export function validateVec3(
  value: RuntimeValue,
  name: string,
  options: { allowMissingZ?: boolean } = {},
): ValidationResult<{ x: number; y: number; z: number }> {
  const objResult = validateObject(value, name);
  if (!objResult.ok) return objResult;

  const obj = objResult.value;
  const xValue = obj.x as RuntimeValue;
  const x = validateFiniteNumber(xValue, `${name}.x`);
  if (!x.ok) return x;

  const yValue = obj.y as RuntimeValue;
  const y = validateFiniteNumber(yValue, `${name}.y`);
  if (!y.ok) return y;

  let z: number;
  if (obj.z === undefined && options.allowMissingZ) {
    z = 0;
  } else {
    const zValue = obj.z as RuntimeValue;
    const zResult = validateFiniteNumber(zValue, `${name}.z`);
    if (!zResult.ok) return zResult;
    z = zResult.value;
  }

  return ok({ x: x.value, y: y.value, z });
}

/**
 * Validate a color object (r, g, b, optional a)
 */
export function validateColor(
  value: RuntimeValue,
  name: string,
): ValidationResult<{ r: number; g: number; b: number; a?: number }> {
  const objResult = validateObject(value, name);
  if (!objResult.ok) return objResult;

  const obj = objResult.value;
  const rValue = obj.r as RuntimeValue;
  const r = validateFiniteNumber(rValue, `${name}.r`, { min: 0, max: 255 });
  if (!r.ok) return r;

  const gValue = obj.g as RuntimeValue;
  const g = validateFiniteNumber(gValue, `${name}.g`, { min: 0, max: 255 });
  if (!g.ok) return g;

  const bValue = obj.b as RuntimeValue;
  const b = validateFiniteNumber(bValue, `${name}.b`, { min: 0, max: 255 });
  if (!b.ok) return b;

  if (obj.a !== undefined) {
    const aValue = obj.a as RuntimeValue;
    const a = validateFiniteNumber(aValue, `${name}.a`, { min: 0, max: 1 });
    if (!a.ok) return a;
    return ok({ r: r.value, g: g.value, b: b.value, a: a.value });
  }

  return ok({ r: r.value, g: g.value, b: b.value });
}

// ════════════════════════════════════════════════════════════════════════════
//                                                    // optional // validators
// ════════════════════════════════════════════════════════════════════════════

/**
 * Make a validator optional - returns undefined if value is undefined/null
 */
export function optional<T extends JSONValue>(
  validator: (value: RuntimeValue, name: string) => ValidationResult<T>,
): (value: RuntimeValue, name: string) => ValidationResult<T | undefined> {
  return (value, name) => {
    if (value === undefined || value === null) {
      return ok(undefined);
    }
    return validator(value, name);
  };
}

/**
 * Provide a default value if undefined
 */
export function withDefault<T extends JSONValue>(
  validator: (value: RuntimeValue, name: string) => ValidationResult<T>,
  defaultValue: T,
): (value: RuntimeValue, name: string) => ValidationResult<T> {
  return (value, name) => {
    if (value === undefined || value === null) {
      return ok(defaultValue);
    }
    return validator(value, name);
  };
}

// ════════════════════════════════════════════════════════════════════════════
//                                                    // composition // helpers
// ════════════════════════════════════════════════════════════════════════════

/**
 * Validate all fields of an object schema
 * Returns typed object if all validations pass
 */
export function validateSchema<T extends ValidatedObject>(
  value: RuntimeValue,
  name: string,
  schema: {
    [K in keyof T]: (value: JSONValue, name: string) => ValidationResult<T[K]>;
  },
): ValidationResult<T> {
  const objResult = validateObject(value, name);
  if (!objResult.ok) return objResult;

  const obj = objResult.value;
  const result: Partial<T> = {};

  for (const key of Object.keys(schema) as (keyof T)[]) {
    const fieldValidator = schema[key];
    const fieldValue = obj[key as string] as JSONValue;
    const fieldResult = fieldValidator(
      fieldValue,
      `${name}.${String(key)}`,
    );
    if (!fieldResult.ok) {
      return fieldResult as ValidationResult<T>;
    }
    result[key] = fieldResult.value;
  }

  return ok(result as T);
}

// ════════════════════════════════════════════════════════════════════════════
//                                                      // error // aggregation
// ════════════════════════════════════════════════════════════════════════════

/**
 * Collect all validation errors instead of failing on first
 */
export function validateAll<T>(
  validations: Array<() => ValidationResult<unknown>>,
  onSuccess: () => T,
): ValidationResult<T> {
  const errors: string[] = [];

  for (const validate of validations) {
    const result = validate();
    if (!result.ok) {
      errors.push(result.error);
    }
  }

  if (errors.length > 0) {
    return fail(errors.join("; "));
  }

  return ok(onSuccess());
}
